_posts/2022-08-17-using-the-kani-rust-verifier-on-tokio-bytes.md

```diff
@@ -0,0 +1,556 @@
+The aim of Tokio is to provide the _"building blocks needed for writing network applications [with] the flexibility to target a wide range of systems, from large servers with dozens of cores to small embedded devices"_.
+In this post we will focus on a low-level component of the Tokio stack and proving properties about a core data structure.
+
+  - [Diving into `Bytes::BytesMut`](#diving-into-bytesbytesmut)
+  - [Using Kani](#using-kani)
+    - [A simple example](#a-simple-example)
+    - [Representation invariants and ghost state](#representation-invariants-and-ghost-state)
+    - [A proof about `with_capacity()`](#a-proof-about-with_capacity)
+    - [A proof about `split_off`](#a-proof-about-split_off)
+    - [About specifications](#about-specifications)
+  - [Summary](#summary)
+
+## Diving into `Bytes::BytesMut`
+
+At the bottom of the Tokio stack is the [Bytes](https://github.com/tokio-rs/bytes) library, a _"rich set of utilities for manipulating byte arrays"_, which reflects the fact that networking applications, at their core, have to manipulate byte streams.
+Inside the library are two container types:
+- `Bytes`: a cheaply cloneable and sliceable chunk of contiguous memory
+- `BytesMut`: a unique reference to a (potentially shared) contiguous slice of memory
+  
+As the names suggest, the bytes pointed to by a `Bytes` instance are immutable whereas they are mutable through a `BytesMut` instance.
+In this post we'll focus on `BytesMut`.
+Let's start with its definition:
+
+```rust
+pub struct BytesMut {
+    ptr: NonNull<u8>,
+    len: usize,
+    cap: usize,
+    data: *mut Shared,
+}
+
+struct Shared {
+    vec: Vec<u8>,
+    original_capacity_repr: usize,
+    ref_count: AtomicUsize,
+}
+```
+
+The straightforward part is the first three fields.
+Similar to a `Vec<u8>`, we have a pointer `ptr` to the backing buffer of size `cap` bytes with `len` bytes currently in use.
+The field `data` is a bit trickier because it handles sharing of the backing buffer.
+In particular, `data` is used to distinguish between two kinds of representation: `KIND_VEC` and `KIND_ARC`.
+- A `KIND_VEC` is used when the backing buffer is only pointed to by `ptr`.
+In this case, `data` is not used as a pointer but instead contains a sentinel value (the bottom bit is set to `'0b1`) to mark that we're using this representation.
+- A `KIND_ARC` is used when the backing buffer is shared by multiple `BytesMut` instances.
+In this case, each instance's `data` field points to a single `Shared` object whose `ref_count` keeps track of the number of aliases.
+In order for this aliasing to be safe, the implementation must ensure that each instance `ptr` field points to a disjoint part of the slice.
+[Additionally, the rules for [pointer alignment](https://doc.rust-lang.org/reference/type-layout.html#pointers-and-references-layout) ensure that we will never mix up representations (since the bottom bit of a pointer will always be `0`).]
```

Comment:
This confused me at first, since a pointer could be pointing to a `u8` in which case its bottom bit need not be `0`. I think you mean to say the bottom bit of a pointer to the `Shared` object cannot be 0, since shared has `align > 1`.

Reply:
We can perhaps add:
```
and the fact that this pointer has an alignment > 1
```
after "Additionally, the rules for pointer alignment" to make it clear.

Reply:
I removed "(since the bottom bit of a pointer will always be `0`)" and added ",and the fact that `Shared` has `align > 1`, " after "pointer alignment". We can add back the sentence between parentheses if you that will make it clearer.